parsing/parser.mly

```diff
@@ -1633,19 +1642,46 @@ let_binding_body:
           | _, Some t -> t
           | _ -> assert false
         in
-        (ghpat(Ppat_constraint(v, ghtyp(Ptyp_poly([],t)))),
+        let typ = ghtyp ~loc:t.ptyp_loc (Ptyp_poly([],t)) in
+        let patloc =
+          { loc_start = rhs_start_pos 1
+          ; loc_end = rhs_end_pos 2
+          ; loc_ghost = true }
+        in
+        (ghpat ~loc:patloc (Ppat_constraint(v, typ)),
          mkexp_constraint $4 $2) }
   | val_ident COLON typevar_list DOT core_type EQUAL seq_expr
-      { (ghpat(Ppat_constraint(mkpatvar $1 1,
-                               ghtyp(Ptyp_poly(List.rev $3,$5)))),
+      { let typloc =
+          { loc_start = rhs_start_pos 3
+          ; loc_end = rhs_end_pos 5
+          ; loc_ghost = true }
+        in
+        let patloc =
+          { loc_start = rhs_start_pos 1
+          ; loc_end = rhs_end_pos 5
+          ; loc_ghost = true }
+        in
```

Comment:
This level of expliciteness works for me, but note that a `rhs_interval` helper that would take two numbers and not set the `ghost` boolean would also work well here -- given that the ghost aspect is set by the `gh*` functions below anyway.

But do these really need to be ghost locations here? Is the idea that `ident : a b . ty` is not a syntactically valid pattern, so this should be ghost? Because intuitively it seems a valid location for a "pattern" that maps to a real part of the program that the user wrote. Same for typloc, which seems reasonable to me.

(I have forgotten what is the implicit contract of non-ghost locations.)

Reply:
I added `rhs_interval` alongside `rhs_loc` in `Location`.

As for the ghosting I have some opinion on the matter, but I'm unwilling to take any action at this point. I'd rather keep this PR simpler and have a look at that at a later point.
As was mentioned in some email, there is a plan of adding a `pvb_type` field to `Parsetree.value_binding` instead of synthesizing constraints on the pattern and sometimes the expression.

Reply:
Even if you want to keep these locations ghost in the produced AST for simplicity, I was pointing out that I think the `with loc_ghost = true` part (in the new patchset) is useless, because the ~loc argument of ghtyp/ghpat is going to get ghosted within the function anyway.

Reply:
Ah indeed! I'll simplify that.